File: src/main/java/org/goafabric/personservice/repository/extensions/TenantResolver.java
package org.goafabric.personservice.repository.extensions;

import io.quarkus.hibernate.orm.PersistenceUnitExtension;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.context.RequestScoped;
import jakarta.enterprise.inject.spi.CDI;
import jakarta.inject.Inject;
import org.eclipse.microprofile.config.ConfigProvider;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.flywaydb.core.Flyway;
import org.goafabric.personservice.extensions.HttpInterceptor;

import java.util.Arrays;
import java.util.Map;

@PersistenceUnitExtension
@RequestScoped
public class TenantResolver implements io.quarkus.hibernate.orm.runtime.tenant.TenantResolver {
    @Inject FlywayConfig flywayConfig;

    @ConfigProperty(name = "multi-tenancy.schema-prefix") String schemaPrefix;

    @Override
    public String getDefaultTenantId() {
        return "PUBLIC";
    }

    @Override
    public String resolveTenantId() {
        return schemaPrefix + HttpInterceptor.getTenantId();
    }

    @ApplicationScoped
    static class FlywayConfig {
        public FlywayConfig() {
            if (ConfigProvider.getConfig().getValue("multi-tenancy.migration.enabled", Boolean.class)) {
                final Flyway flyway = CDI.current().select(Flyway.class).get();
                final String schemas = ConfigProvider.getConfig().getValue("multi-tenancy.tenants", String.class);
                final String schemaPrefix = ConfigProvider.getConfig().getValue("multi-tenancy.schema-prefix", String.class);
                Arrays.asList(schemas.split(",")).forEach(schema -> {
                            Flyway.configure()
                                    .configuration(flyway.getConfiguration())
                                    .schemas(schemaPrefix + schema)
                                    .defaultSchema(schemaPrefix + schema)
                                    .placeholders(Map.of("tenantId", schema))
                                    .load()
                                    .migrate();
                        }
                );
            }
        }
    }


}

File: doc/examples/discriminator/TenantResolver.kt
package org.goafabric.personservice.persistence.extensions

import io.quarkus.hibernate.orm.PersistenceUnitExtension
import io.quarkus.runtime.StartupEvent
import jakarta.annotation.Priority
import jakarta.enterprise.context.RequestScoped
import jakarta.enterprise.event.Observes
import jakarta.enterprise.inject.spi.CDI
import jakarta.inject.Inject
import org.eclipse.microprofile.config.ConfigProvider
import org.eclipse.microprofile.config.inject.ConfigProperty
import org.flywaydb.core.Flyway
import org.goafabric.personservice.extensions.UserContext.tenantId
import org.goafabric.personservice.persistence.DemoDataImporter

@PersistenceUnitExtension
@RequestScoped
class TenantResolver: io.quarkus.hibernate.orm.runtime.tenant.TenantResolver {

    @ConfigProperty(name = "multi-tenancy.schema-prefix")
    var schemaPrefix: String? = null

    @Inject
    lateinit var demoDataImporter: DemoDataImporter

    override fun getDefaultTenantId(): String {
        return tenantId
    }

    override fun resolveTenantId(): String {
        return tenantId
    }

    fun onStart(@Observes @Priority(1) ev: StartupEvent) {
        if (ConfigProvider.getConfig().getValue("database.provisioning.goals", String::class.java).contains("-migrate")) {
            val flyway = CDI.current().select(Flyway::class.java).get()
            Flyway.configure()
                .configuration(flyway.configuration)
                .load()
                .migrate()
        }
    }
}
